### PR TITLE
Add concurrent option

### PR DESCRIPTION
**Pull request by brendt**

### CODE REVIEW

File: tests/ForkTest.php [hunk 1]
```diff
@@ -29,6 +30,33 @@ public function it_will_execute_the_given_closures()
         $this->assertEquals([2, 4], $results);
     }
 
+    /** @test */
+    public function it_will_execute_the_given_closures_with_concurrency_cap()
```
Comment (freekmurze) [inline tests/ForkTest.php:34]: Nice!

File: src/Fork.php [hunk 2]
```diff
@@ -80,26 +85,66 @@ protected function forkForTask(Task $task): Task
...
+        $amountRunning = 0;
+
+        foreach ($queue as $i => $task) {
+            $running[] = $this->runTask($task);
+
+            unset($queue[$i]);
+
+            $amountRunning += 1;
+
+            if ($this->concurrent && $amountRunning >= $this->concurrent) {
```
Comment (freekmurze) [inline src/Fork.php:123]: Let's extract this to a function with a good description, `maximumAmountOfTasksRunning` or better.

File: src/Fork.php [hunk 3]
```diff
@@ -80,26 +85,66 @@ protected function forkForTask(Task $task): Task
...
+                    continue;
+                }
+
+                $output[$task->order()] = $this->finishTask($task);
+
+                unset($running[$key]);
 
-                    $output[$task->order()] = $taskOutput;
+                if (count($queue)) {
+                    $i = array_key_first($queue);
```
Comment (freekmurze) [inline src/Fork.php:139]: Nice!

File: src/Fork.php [hunk 4]
```diff
@@ -80,26 +85,66 @@ protected function forkForTask(Task $task): Task
...
 
-        while (count($runningTasks)) {
-            foreach ($runningTasks as $key => $task) {
-                if ($task->isFinished()) {
-                    $taskOutput = $task->output();
+        $running = [];
+
+        $amountRunning = 0;
+
+        foreach ($queue as $i => $task) {
```
Comment (freekmurze) [inline src/Fork.php:116]: Maybe we could extract this loop to startInitialTasks

File: src/Fork.php [hunk 5]
```diff
@@ -80,26 +85,66 @@ protected function forkForTask(Task $task): Task
...
+
+        while (count($running)) {
+            foreach ($running as $key => $task) {
+                if (! $task->isFinished()) {
+                    continue;
+                }
+
+                $output[$task->order()] = $this->finishTask($task);
+
+                unset($running[$key]);
```
Comment (freekmurze) [inline src/Fork.php:136]: Maybe we could extract the remainder to `startNewTask`

**Review by freekmurze — CHANGES_REQUESTED** [review]

**Review by freekmurze — CHANGES_REQUESTED** [review]
Looks nice.


We could opt to extract some things in  `waitFor` to their own functions so it's more clear what it happing there. If it would make it easier, I wouldn't made setting `$running` as an instance variable.

**Comment by brendt** [comment]
PHP cs is acting up, but I'm going to merge it nevertheless.

**Comment by brendt** [comment]
wrong button ><

**Pull request merged**